src/main/java/com/limechain/trie/structure/Vacant.java

```diff
@@ -0,0 +1,482 @@
+            private final TrieNode.Parent branchParent;
+
+            /**
+             * Value of {@link TrieNode#partialKey} for the newly-created branch node.
+             */
+            private final Nibbles branchPartialKey;
+
+            /**
+             * Value of {@link TrieNode#childrenIndices} for the newly-created branch node.
+             * Does not include the entry that must be filled with the new storage node.
+             */
+            private final Integer[] branchChildrenIndices;
+
+
+            private Two(TrieStructure<T> trieStructure,
+                        Nibble storageChildIndex,
+                        Nibbles storagePartialKey,
+                        @Nullable TrieNode.Parent branchParent,
+                        Nibbles branchPartialKey,
+                        Integer[] branchChildrenIndices) {
+                super(trieStructure);
+                this.storageChildIndex = storageChildIndex;
+                this.storagePartialKey = storagePartialKey;
+                this.branchParent = branchParent;
+                this.branchPartialKey = branchPartialKey;
+                this.branchChildrenIndices = branchChildrenIndices;
+            }
+
+
+            /**
+             * Insert the new storage node (and the intermediate branch node needed for this case)
+             * Branch node's user data defaults to null.
+             * @param storageUserData the userData held by the storage node
+             * @return a StorageNodeHandle to the freshly inserted storage node
+             */
+            public StorageNodeHandle<T> insert(T storageUserData) {
+                return this.insert(storageUserData, null);
+            }
+
+            /**
+             * Insert the new storage node (and the intermediate branch node needed for this case)
+             * @param storageUserData the userData held by the storage node
+             * @param branchUserData the userData held by the branch node
+             * @return a StorageNodeHandle to the freshly inserted storage node
+             */
+            @SuppressWarnings("unused")
+            public StorageNodeHandle<T> insert(T storageUserData, T branchUserData) {
+                int newBranchNodePartialKeyLen = this.branchPartialKey.size();
+
+                assert 1 == Arrays.stream(this.branchChildrenIndices).filter(Objects::nonNull).count()
+                    : "The branch node we're about to insert must have exactly one child (the node that previously existed before this insertion)";
+
+                // Insert the intermediate branch node
+                int newBranchNodeIndex = this.trieStructure.nodes.add(new TrieNode<>(
+                    this.branchParent,
+                    this.branchPartialKey,
+                    // NOTE: Crucial mutability moment, this has to be a copy of the array.
+                    //  I'll never get those three hours of my life back :)
+                    //  Manage your ownership, kids.
```

Comment:
I would remove this note, but I really want to emphasize the utmost importance of cloning the array here, since I had a very turbulent 2-3 hours debugging mutability issues since I had forgotten to clone it and was updating the same reference a few lines below. I'm open for suggestions on how to better annotate it, with no comment it appears totally random and inconspicuous.

Reply:
What reference is being updated? `branchChildrenIndices`  seems to stay the same throughout.

Reply:
The variable itself doesn't, but exactly in this line we're passing it as the children indices for the new branch node, crucially, without having the new storage node inserted. That's why copying the old snapshot is very important, since two lines below we're also adding the freshly created storage node to its parent's children indices array:
https://github.com/LimeChain/Fruzhin/blob/0b505ce950e737ce64882c305834b045cab27045/src/main/java/com/limechain/trie/structure/Vacant.java#L459-L461

Thus, if we didn't copy, this mutates the original variable as well, which in turn breaks the loop where we're updating its **old** children:
https://github.com/LimeChain/Fruzhin/blob/0b505ce950e737ce64882c305834b045cab27045/src/main/java/com/limechain/trie/structure/Vacant.java#L463

The crucial difference between the old children and this new storage node child is their partial keys, which got messed up - a single character missing in seemingly random places, which was the result of the above-mentioned lack of copying.